openvdb_points/tools/IndexIterator.h

```diff
@@ -0,0 +1,272 @@
+
+    inline Index32 end() { return mIndexIter.end(); }
+
+    inline void reset(Index32 item, Index32 end) {
+        mIndexIter.reset(item, end);
+    }
+
+    /// @brief  Returns the item to which this iterator is currently pointing.
+    inline Index32 operator*() { return *mIndexIter; }
+    inline Index32 operator*() const { return *mIndexIter; }
+
+    /// @brief  Return @c true if this iterator is not yet exhausted.
+    inline operator bool() const { return mIter; }
+    inline bool test() const { return mIter; }
+
+    /// @brief  Advance to the next (valid) item (prefix).
+    inline IndexValueIter& operator++() {
+        mIndexIter.next();
+        while (!mIndexIter.test() && mIter.next()) {
+            mIndexIter.reset(mParent.getValue(mIter.offset() - 1), *mIter);
+        }
+        return *this;
+    }
+
+    /// @brief  Advance to the next (valid) item (postfix).
+    inline IndexValueIter operator++(int /*dummy*/) {
+        IndexIter newIterator(*this);
+        this->operator++();
+        return newIterator;
+    }
+
+    /// @brief  Advance to the next (valid) item.
+    inline bool next() { this->operator++(); return this->test(); }
+    inline bool increment() { this->next(); return this->test(); }
+
+    /// @brief Equality operators
+    bool operator==(const IndexValueIter& other) const { return *mIndexIter == *other.mIndexIter; }
+    bool operator!=(const IndexValueIter& other) const { return !this->operator==(other); }
+
+private:
+    IndexIter mIndexIter;
+    ValueIterT mIter;
+    const typename ValueIterT::NodeType& mParent;
+}; // IndexValueIter
+
+
+/// IndexIterTraits provides the following for iterators of the three value
+/// types, i.e., for {Value}{On,Off,All}{CIter}:
+/// - a begin(leaf) function that returns an index iterator or an index value
+///   iterator for the leaf provided,
+///   eg IndexIterTraits<Tree, Tree::LeafNodeType::ValueOn>::begin(leaf) returns
+///   leaf.beginIndexOn()
+/// - an Iterator typedef that aliases to the index iterator for this value type
+template<typename TreeT, typename ValueT> struct IndexIterTraits;
+
+template<typename TreeT>
+struct IndexIterTraits<TreeT, typename TreeT::LeafNodeType::ValueAllCIter> {
+    typedef IndexIter Iterator;
+    static Iterator begin(const typename TreeT::LeafNodeType& leaf) {
+        return Iterator(leaf.beginIndexAll());
+    }
+};
+
+template<typename TreeT>
+struct IndexIterTraits<TreeT, typename TreeT::LeafNodeType::ValueOnCIter> {
+    typedef typename TreeT::LeafNodeType::IndexOnIter Iterator;
+    static Iterator begin(const typename TreeT::LeafNodeType& leaf) {
+        return Iterator(leaf.beginIndexOn());
+    }
+};
+
+template<typename TreeT>
+struct IndexIterTraits<TreeT, typename TreeT::LeafNodeType::ValueOffCIter> {
+    typedef typename TreeT::LeafNodeType::IndexOffIter Iterator;
+    static Iterator begin(const typename TreeT::LeafNodeType& leaf) {
+        return Iterator(leaf.beginIndexOff());
+    }
+};
+
+
```

Comment:
I think the iterators are quite nice! But the FilterIndexIter could definitely use a comment here with the expected implementation of the FilterT and example usage.